Strict safety improvements

DougGregor opened the pull request:
Collected improvements to the strict memory-safety mode to better align with the proposal:
* Diagnose `@preconcurrency import`, since it admits data races that can undermine memory safety
* Diagnose the use of `-Ounchecked`, which disables safety checks
* Introduce `@unsafe` conformances, diagnose when we need them (unsafe witnesses for safe requirements, collated nicely), and diagnose when using them
* Properly account for the presence of unsafe types in declarations that aren't themselves marked `@unsafe` (e.g., because they're in a module that doesn't use strict safety) in type aliases (https://github.com/swiftlang/swift/issues/78220), witnessed requirements, and overridden declarations.
* Emit loaded module trace information about modules and dependencies that enabled strict memory safety checking

This third one required a significant refactoring of how we keep track of the attributes that apply to a conformance, like `@unchecked`, `@preconcurrency`, and so on. That's the bulk of the code change here.

Fixes rdar://127128995

DougGregor committed:
Migrate "@unchecked Sendable" strict safety diagnostic to become a note associated with its type

DougGregor committed:
Diagnose uses of `@unchecked Sendable` conformances, not declarations

Under strict concurrency and memory safety, uses of `@unchecked
Sendable` conformances are considered unsafe. Diagnose the use sites,
not the declaration site.

DougGregor committed:
Diagnose @preconcurrency imports as a strict safety issue

@preconcurrency imports disable Sendable checking, which can lead to
data races that undermine memory safety. Diagnose such imports, and
require `@safe(unchecked)` to suppress the diagnostic.

DougGregor committed:
Diagnose the combination of -Ounchecked with strict memory safety

Aligns with the current proposal and fixes rdar://127128995

DougGregor committed:
Allow `@safe(unchecked)` on import decls

DougGregor committed:
Remove WarnUnsafe from the standard library build until we're ready to deal with the warnings

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Only emit @unsafe warning when strict safety mode is enabled

DougGregor commented:
@swift-ci please smoke test

Azoy commented:
> Diagnose uses of @unchecked Sendable conformances

I'm confused, is using `Mutex` a safety violation here? I don't think this makes sense.

DougGregor commented:
> I'm confused, is using `Array`, `Dictionary`, or `Mutex` a safety violation here? I don't think this makes sense, this should be the same principle of `@safe(unchecked)` where uses of it as considered _safe_.

Ah, you're right, of course. We've already indicated that it can't be checked at the definition site. Strict memory safety checking doesn't need to diagnose either of these. I'll revert that bit and remove it from the proposal, thanks!

DougGregor committed:
Generalize protocol conformance options spelled via attribute and incorporate @unsafe

Protocol conformances have a handful attributes that can apply to them
directly, including @unchecked (for Sendable), @preconcurrency, and
@retroactive. Generalize this into an option set that we carry around,
so it's a bit easier to add them, as well as reworking the
serialization logic to deal with an arbitrary number of such options.

Use this generality to add support for @unsafe conformances, which are
needed when unsafe witnesses are used to conform to safe requirements.
Implement general support for @unsafe conformances, including
producing a single diagnostic per missing @unsafe that provides a
Fix-It and collects together all of the unsafe witnesses as notes.

DougGregor committed:
Mark SerialExecutor.asUnownedSerialExecutor unsafe so conformances don't need to be @unsafe

What we should actually do with SerialExecutor conformances is not yet clear, though.

DougGregor committed:
Don't diagnose uses of `@unchecked Sendable` conformances

The `@unchecked` conformance is effectively the same as
`@safe(unchecked)`, in that it asserts memory safety in a place where
it cannot be automatically checked. But once that has been asserted,
there is no reason to diagnose anywhere else.

While here, drop the "unsafe declaration here" note, which isn't
adding value but did add noise.

Thanks, Alex!

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor commented:
@swift-ci please clean build toolchain macOS

DougGregor commented:
@swift-ci please clean build toolchain macOS

DougGregor commented:
@swift-ci please clean build toolchain macOS

DougGregor commented:
@swift-ci please clean build toolchain macOS

DougGregor commented:
@swift-ci please build toolchain macOS

DougGregor committed:
Diagnose the use of a typealias that involves unsafe types

As we do when referencing other kinds of declarations, if a
typealias isn't `@unsafe`, but it involves unsafe types,
diagnose the non-safety at the point of reference.
Fixes https://github.com/swiftlang/swift/issues/78220

DougGregor committed:
Account for implicit unsafety in unsafe witness and override checks

When a witness is unsafe and its corresponding requirement is not
unsafe, the conformance itself needs to be determined to be unsafe.
When doing this check, account for the fact that the requirement might
be implicitly unsafe, i.e., it uses unsafe types. In such cases, the
requirement is effectively unsafe, so the unsafe witness to it does not
make the conformance unsafe. Therefore, suppress the diagnostic. This
accounts for cases where the protocol comes from a module that didn't
enable strict memory safety checking, or didn't suppress all warnings
related to it, as well as cases where substitution of type witnesses
introduces the unsafe type.

The same thing occurs with overriding a method: an unsafe override of
a method that involves unsafe types (but was not marked @unsafe for
whatever reason) does not make the subclassing itself unsafe, so don't
diagnose it as such.

DougGregor committed:
Factor the strict memory safety checking declarations into their own header

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor committed:
Add module trace information for strict memory safety

Extend the module trace format with a field indicating whether a given
module, or any module it depends on, was compiled with strict memory
safety enabled. This separate output from the compiler can be used as
part of an audit to determine what parts of Swift programs are built
with strict memory safety checking enabled.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor committed:
Fix loaded module trace tests

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor committed:
Fix substitutions for generic requirements

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

Xazax-hun commented on include/swift/AST/DiagnosticsFrontend.def:
```diff
@@ -589,5 +589,9 @@ ERROR(experimental_not_supported_in_production,none,
       (StringRef))
 
+GROUPED_WARNING(Ounchecked_with_strict_safety,Unsafe,none,
+      "'-Ounchecked' is not memory-safe and should not be combined with "
```
I wonder what do you think about `-disable-access-control`. It might be used to circumvent some safety checks, but it is a frontend only flag. Should we also diagnose something like that?
